### ISSUE TITLE
IAM policy array with single item is dirty in future plans.

### ISSUE THREAD

**Issue by dcosson** [hashicorp/terraform#8618]
AWS (and terraform) will let you successfully submit an IAM policy with a single item in an array e.g. `"Action": ["*"]`, but then that value will show up as dirty in future `terraform plan`s because AWS returns the policy document without the array e.g. `"Action": "*"`.

I've run into this a few times now, and it would be nicer if it failed fast instead. Would it be possible for terraform to do an extra validation step in the plan stage on an IAM policy json document and then give a descriptive error if you submit a document like this?

**Comment by stack72** [comment]
Hi @dcosson 

I wish i had better news than we know this to be an issue with IAM Policies - fortunately, we are going to be fixing this real soon! We are going to be able to perform structural analysis on the Policy objects so keep tuned :)

P.

**Comment by stack72** [comment]
Hi @dcosson 

Ok, so I am pleased to say that #8585 will solve the issue you are having. In a nutshell, the library that fixes the issue will treat this:

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Effect": "Allow",
      "Principal": {
        "Service": "spotfleet.amazonaws.com"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
```

the same as this

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Action": ["sts:AssumeRole"],
      "Effect": "Allow",
      "Principal": {
        "Service": "spotfleet.amazonaws.com"
      }
    }
  ]
}
```

This will be released as part of Terraform 0.7.3 :)

Paul
